File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;
class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int answer = 0;
        
        int lt = 1;
        int rt = Arrays.stream(diffs).max().getAsInt() + 1;
        
        while (lt <= rt ){
            int mid = (lt+rt)/2;
            long l = limit;
            
            for(int i=0; i<diffs.length;i++){
                if(diffs[i]<=mid){
                    l -= times[i];
                }else{
                    int cur = diffs[i] - mid;
                    int prev = (i == 0) ? 0 : times[i-1];
                    l -= (long)(prev+times[i])*cur;
                    l -= times[i];
                }
            }
            if(l<0){
                lt = mid +1;
            }else{
                rt = mid-1;
            }
            
        }
            
        return lt;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = 0
    n = len(times)

    lt = 1
    rt = max(diffs)

    while lt<rt:
        mid = (lt+rt)//2
        
        cur =0 
        for i in range(n):
            if(mid >= diffs[i]):
                cur += times[i]
            else:
                if(i<=0):
                    cur += times[i]*(diffs[i]-mid)+times[i]
                else:
                    cur += (times[i]+times[i-1])*(diffs[i]-mid)+times[i]
        
        if(cur<=limit):
            rt = mid
        else:
            lt = mid+1
            

    return lt